File: Services/FfmpegConfig.cs
using System.IO;

namespace x264video4osu.Services;

/// <summary>
/// FFmpeg 配置类
/// </summary>
public static class FfmpegConfig
{
    private const string FfmpegDownloadUrl = "https://www.gyan.dev/ffmpeg/builds/ffmpeg-release-essentials.zip";

    /// <summary>
    /// FFmpeg 可执行文件路径
    /// </summary>
    public static string FfmpegPath => GetToolPath("ffmpeg.exe");

    /// <summary>
    /// FFprobe 可执行文件路径
    /// </summary>
    public static string FfprobePath => GetToolPath("ffprobe.exe");

    /// <summary>
    /// FFmpeg 下载地址
    /// </summary>
    public static string DownloadUrl => FfmpegDownloadUrl;

    /// <summary>
    /// Tools 文件夹路径
    /// </summary>
    public static string ToolsFolder => Path.Combine(AppDomain.CurrentDomain.BaseDirectory, "tools");

    /// <summary>
    /// 验证 FFmpeg 工具是否存在，不存在则抛出异常
    /// </summary>
    public static void ValidateTools()
    {
        string missingTools = string.Empty;

        if (!File.Exists(FfmpegPath))
            missingTools += "• ffmpeg.exe\n";

        if (!File.Exists(FfprobePath))
            missingTools += "• ffprobe.exe\n";

        if (!string.IsNullOrEmpty(missingTools))
        {
            throw new FileNotFoundException(missingTools, ToolsFolder);
        }
    }

    /// <summary>
    /// 获取工具路径，只从本地 tools 文件夹获取，不存在则抛出异常
    /// </summary>
    private static string GetToolPath(string exeName)
    {
        // 优先使用项目根目录下的 tools 文件夹
        string toolsPath = Path.Combine(AppDomain.CurrentDomain.BaseDirectory, "tools", exeName);
        if (File.Exists(toolsPath))
            return toolsPath;

        // 尝试使用相对路径 tools 文件夹
        string relativeToolsPath = Path.Combine("tools", exeName);
        if (File.Exists(relativeToolsPath))
            return relativeToolsPath;

        // 不存在时返回路径（由调用者检查）
        return toolsPath;
    }
}

File: src/services/ffmpeg_config.rs
//! FFmpeg / ffprobe 工具定位与校验（对应旧程序 Services/FfmpegConfig.cs）。

use crate::error::{AppError, AppResult};
use std::path::PathBuf;

const FFMPEG_DOWNLOAD_URL: &str = "https://www.gyan.dev/ffmpeg/builds/ffmpeg-release-essentials.zip";

/// 应用可执行文件所在目录
fn exe_dir() -> PathBuf {
    std::env::current_exe()
        .ok()
        .and_then(|p| p.parent().map(|p| p.to_path_buf()))
        .unwrap_or_default()
}

/// 与应用同级的 tools 文件夹（对应 `AppDomain.CurrentDomain.BaseDirectory + "tools"`）
pub fn tools_folder() -> PathBuf {
    exe_dir().join("tools")
}

/// 优先 `<exe>/tools/<exeName>`，其次相对 cwd 的 `tools/<exeName>`，
/// 都不存在时返回前者（由调用者检查）。对应 `GetToolPath`。
fn get_tool_path(exe_name: &str) -> PathBuf {
    let primary = exe_dir().join("tools").join(exe_name);
    if primary.exists() {
        return primary;
    }

    let relative = PathBuf::from("tools").join(exe_name);
    if relative.exists() {
        return relative;
    }

    primary
}

pub fn ffmpeg_path() -> PathBuf {
    get_tool_path("ffmpeg.exe")
}

pub fn ffprobe_path() -> PathBuf {
    get_tool_path("ffprobe.exe")
}

pub fn download_url() -> &'static str {
    FFMPEG_DOWNLOAD_URL
}

/// 校验工具是否存在，缺失时返回带缺失清单的错误（对应 `ValidateTools` 抛
/// `FileNotFoundException`，消息包含 `• ffmpeg.exe\n`）。
pub fn validate_tools() -> AppResult<()> {
    let mut missing = String::new();
    if !ffmpeg_path().exists() {
        missing.push_str("• ffmpeg.exe\n");
    }
    if !ffprobe_path().exists() {
        missing.push_str("• ffprobe.exe\n");
    }
    if !missing.is_empty() {
        return Err(AppError::ToolsMissing(missing));
    }
    Ok(())
}
